### PR TITLE
Add neopixel (ws2812 and APA106) support for esp32

### PR DESCRIPTION
**Pull request by ghost**
I believe this follows guidelines, but feedback welcome.
 
This basically ports the esp8266 neopixel code. The main changes were in neopixel.c, to adjust to the new IRQ and CPU frequency stuff. I believe I've done that the right way, by using the ESP32 code for quiet_time, etc, but it involved also copying xirq stuff across - the IRQ macros for enable and disable seemed to be defined but not implemented. Happy to revisit it, obviously, but I'd need a little more guidance in how to do it the ESP32 way if this is not good. 

The main difference between APA106 and the existing ws2812 seems to be the RGB order, so I've added a simple AP106 python module to wrap the noepixel class with a different order.

It works with the latest esp32 branch as of this morning, and that uses the latest supported esp-idf.

### CODE REVIEW

**Comment by hoihu** [comment]
It may be better to use the RMT module - The HW timing is then provided by this module rather than bitbanging (I havent looked into your code but I recall that the esp8266 port did it that way)

There are various pointers in the esp32 forum on how to do this.

Pycom will merge its fork soon (which includes the RMT version of its neopixel driver). I guess that this  will make your pull request obsolete I'm afraid.

As for style of this patch, you should rebase and rephrase the commit message.

**Comment by dpgeorge** [comment]
> It may be better to use the RMT module

Does the RMT module allow to route to any IO pin?  And can you use it on multiple pins?  It would be great if you could have ws2812/apa102 on arbitrary pins, and on as many pins at once as you liked.  Also, using the RMT for this would take away the RMT from other uses.

**Comment by hoihu** [comment]
I checked chapter 9 of the docs
http://espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf#page181

I see 8 independant channels that can be configured individually. The only difference in the RMT channels seems to be the way the RAM is mapped - e.g. channel 0 can acces all 512x32 bytes while channel 7 can only access the last page. So for larger waveforms (or neopixel string) channel 0 seems to be the logical pick. At least if you want to stream the data out without CPU intervention. If IRQs are used I believe it shouldn't matter since the IRQ handler would then setup the next data packet.

It's seems to be a fun peripheral module😉, not only for neopixel. But I'm also new to this CPU so its hard to judge design risks. 

As for routing to pins: The according function in the SDK takes any pin number so I guess it is possible to route them to all pins.

Is it better than bitbanging? Maybe only if the data must be streamed without intervention of the CPU. Otherwise, perhaps it's more straightforward to use a SW solution... and correct - it would free the RMT module ( at least one channel of the RMT).

**Comment by dpgeorge** [comment]
Ok, well, the RMT peripheral does indeed look very configurable.  But also some effort to implement.

I think it's fair to merge this PR (with some cleanup) since it's relatively simple and works.  Then the RMT module can perhaps one day replace it.

File: esp32/espneopixel.c [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+#include "py/mpconfig.h"
+#if MICROPY_ESP32_NEOPIXEL
+
+#include "rom/ets_sys.h"
+#include "espneopixel.h"
+#include "mphalport.h"
+
+#define NEO_KHZ400 (1)
+
+void /*ICACHE_RAM_ATTR*/ esp_neopixel_write(uint8_t pin, uint8_t *pixels, uint32_t numBytes, bool is800KHz) {
```
Comment (dpgeorge) [inline esp32/espneopixel.c]: You'll probably need this ICACHE_RAM_ATTR (or the esp32 equivalent).

File: esp32/espneopixel.c [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+  pinMask   = 1 << pin;
+  p         =  pixels;
+  end       =  p + numBytes;
+  pix       = *p++;
+  mask      = 0x80;
+  startTime = 0;
+
+  uint32_t fcpu = ets_get_cpu_frequency() * 1000000;
+// for APA106 - 0.35 us for time0, 1.35 us for time1, inferred period of ~1.75 us
```
Comment (dpgeorge) [inline esp32/espneopixel.c]: Does the APA use different timing?  What is this comment for?

File: esp32/espneopixel.c [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+// Needs to be a separate .c file to enforce ICACHE_RAM_ATTR execution.
+
+#include "py/mpconfig.h"
+#if MICROPY_ESP32_NEOPIXEL
+
+#include "rom/ets_sys.h"
+#include "espneopixel.h"
+#include "mphalport.h"
+
+#define NEO_KHZ400 (1)
```
Comment (dpgeorge) [inline esp32/espneopixel.c]: We can remove this config option (always allow for 400khz timing).

File: esp32/modesp.c [hunk 4]
```diff
@@ -30,7 +30,8 @@
 #include <stdio.h>
 
 #include "esp_spi_flash.h"
-
+#include "espneopixel.h"
+#include "mphalport.h"
```
Comment (dpgeorge) [inline esp32/modesp.c]: Include "py/mphal.h" instead.

File: esp32/modesp.c [hunk 5]
```diff
@@ -78,6 +79,16 @@ STATIC mp_obj_t esp_flash_user_start(void) {
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_0(esp_flash_user_start_obj, esp_flash_user_start);
 
+STATIC mp_obj_t esp_neopixel_write_(mp_obj_t pin, mp_obj_t buf, mp_obj_t is800k) {
```
Comment (dpgeorge) [inline esp32/modesp.c]: Probably better to make the is800k parameter a generic "timing" parameter.  Eg a value of 0 is 400khz, 1 is 800khz, 2 could be APA (if it differs), etc.

File: esp32/moduos.c [hunk 6]
```diff
@@ -37,6 +37,7 @@
 #include "py/runtime.h"
 #include "py/mperrno.h"
 #include "genhdr/mpversion.h"
+//#include "esp_mphal.h"
```
Comment (dpgeorge) [inline esp32/moduos.c]: Please remove this change.

File: esp32/mpconfigport.h [hunk 7]
```diff
@@ -140,6 +140,8 @@
     { MP_OBJ_NEW_QSTR(MP_QSTR_input), (mp_obj_t)&mp_builtin_input_obj }, \
     { MP_OBJ_NEW_QSTR(MP_QSTR_open), (mp_obj_t)&mp_builtin_open_obj },
 
+#define MICROPY_ESP32_NEOPIXEL		(1)
+#define MICROPY_ESP32_APA106		(1)
```
Comment (dpgeorge) [inline esp32/mpconfigport.h]: No need for the APA option, it does nothing.

File: esp32/scripts/lca_backlight.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
+# micropython ESP32 - turn on and off LCD backlight for LCA2017 "badge"
```
Comment (dpgeorge) [inline esp32/scripts/lca_backlight.py]: This script should not be part of this PR.

File: esp32/scripts/lca_neopixel.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
+# micropython ESP32 - turn on and off LCD backlight for LCA2017 "badge"
```
Comment (dpgeorge) [inline esp32/scripts/lca_neopixel.py]: This script shouldn't be part of this PR.

File: esp32/xtirq.h [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#ifndef __MICROPY_INCLUDED_ESP8266_XTIRQ_H__
+#define __MICROPY_INCLUDED_ESP8266_XTIRQ_H__
```
Comment (dpgeorge) [inline esp32/xtirq.h]: All 8266's should be changed to 32 in this file.

Comment (ghost) [reply esp32/xtirq.h]: Hi, sorry, I've been on holiday and starting a new job. I will look at these and clean the code up this weekend - that all makes sense, thanks.

**Comment by dpgeorge** [comment]
@tyggerjai thanks for updating the PR.  I squashed it, fixed a few things and merged it in f3913bfdfb3be3855fd19b351355a7eb51b7dc39
